core/services/vrf/v2/listener_v2_log_listener.go

```diff
@@ -0,0 +1,348 @@
+				// wrong keyhash, can ignore
+				continue
+			}
+			requested[parsed.RequestID().String()] = parsed
+			requestedLP[parsed.RequestID().String()] = l
+		}
+	}
+	// should never happen, unsure if recoverable
+	// may be worth a panic
+	if errs != nil {
+		ll.Errorw("encountered parse errors", "err", errs)
+	}
+
+	if len(fulfilled) > 0 || len(requested) > 0 {
+		ll.Infow("found logs", "fulfilled", len(fulfilled), "requested", len(requested))
+	} else {
+		ll.Debugw("no logs found")
+	}
+
+	// find unfulfilled requests by comparing requested events with the fulfilled events
+	for reqID, req := range requested {
+		if _, isFulfilled := fulfilled[reqID]; !isFulfilled {
+			unfulfilled = append(unfulfilled, req)
+			unfulfilledLP = append(unfulfilledLP, requestedLP[reqID])
+		}
+	}
+
+	return unfulfilled, unfulfilledLP, fulfilled
+}
+
+func (lsn *listenerV2) getConfirmedAt(req RandomWordsRequested, nodeMinConfs uint32) uint64 {
+	// Take the max(nodeMinConfs, requestedConfs + requestedConfsDelay).
+	// Add the requested confs delay if provided in the jobspec so that we avoid an edge case
+	// where the primary and backup VRF v2 nodes submit a proof at the same time.
+	minConfs := nodeMinConfs
+	if uint32(req.MinimumRequestConfirmations())+uint32(lsn.job.VRFSpec.RequestedConfsDelay) > nodeMinConfs {
+		minConfs = uint32(req.MinimumRequestConfirmations()) + uint32(lsn.job.VRFSpec.RequestedConfsDelay)
+	}
+	newConfs := uint64(minConfs) * (1 << lsn.respCount[req.RequestID().String()])
+	// We cap this at 200 because solidity only supports the most recent 256 blocks
+	// in the contract so if it was older than that, fulfillments would start failing
+	// without the blockhash store feeder. We use 200 to give the node plenty of time
+	// to fulfill even on fast chains.
+	if newConfs > 200 {
+		newConfs = 200
+	}
+	if lsn.respCount[req.RequestID().String()] > 0 {
+		lsn.l.Warnw("Duplicate request found after fulfillment, doubling incoming confirmations",
+			"txHash", req.Raw().TxHash,
+			"blockNumber", req.Raw().BlockNumber,
+			"blockHash", req.Raw().BlockHash,
+			"reqID", req.RequestID().String(),
+			"newConfs", newConfs)
+		vrfcommon.IncDupeReqs(lsn.job.Name.ValueOrZero(), lsn.job.ExternalJobID, lsn.coordinator.Version())
+	}
+	return req.Raw().BlockNumber + newConfs
+}
+
+func (lsn *listenerV2) handleFulfilled(fulfilled map[string]RandomWordsFulfilled) {
+	for _, v := range fulfilled {
+		// don't process same log over again
+		// log key includes block number and blockhash, so on re-orgs it would return true
+		// and we would re-process the re-orged request.
+		if !lsn.fulfillmentLogDeduper.ShouldDeliver(v.Raw()) {
+			continue
+		}
+		lsn.l.Debugw("Received fulfilled log", "reqID", v.RequestID(), "success", v.Success())
+		lsn.respCount[v.RequestID().String()]++
+		lsn.blockNumberToReqID.Insert(fulfilledReqV2{
+			blockNumber: v.Raw().BlockNumber,
+			reqID:       v.RequestID().String(),
+		})
```

Comment:
In case of a re-org, should we delete the previous blockNumber entry for the same reqID?